File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/IsValid.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayDeque;

public class IsValid {
    public boolean isValid(String s) {
        ArrayDeque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.addLast(c);
            } else {
                if (stack.isEmpty()) return false;

                char top = stack.removeLast();
                if ((c == ')' && top != '(') ||
                        (c == ']' && top != '[') ||
                        (c == '}' && top != '{')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/isValid.kt
package com.only4.algorithm.version4k.leetcode

fun isValid(s: String): Boolean {
    val stack = ArrayDeque<Char>()

    for (c in s) {
        when (c) {
            '(', '[', '{' -> stack.addLast(c)
            else -> {
                if (stack.isEmpty()) return false

                val top = stack.removeLast()
                when (c) {
                    ')' -> if (top != '(') return false
                    ']' -> if (top != '[') return false
                    '}' -> if (top != '{') return false
                }
            }
        }
    }

    return stack.isEmpty()
}
